modules/s3_bucket/main.tf

```diff
@@ -0,0 +1,90 @@
+resource "aws_s3_bucket" "this" {
+  bucket = var.name
+}
+
+resource "aws_s3_bucket_ownership_controls" "this" {
+  bucket = aws_s3_bucket.this.id
+
+  rule {
+    object_ownership = "BucketOwnerEnforced"
+  }
+}
+
+resource "aws_s3_bucket_public_access_block" "this" {
+  bucket = aws_s3_bucket.this.id
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+resource "aws_s3_bucket_policy" "this" {
+  bucket = aws_s3_bucket.this.id
+  policy = data.aws_iam_policy_document.this.json
+}
+
+data "aws_iam_policy_document" "this" {
+  source_policy_documents = var.bucket_policies
+
+  statement {
+    sid    = "RequireHttps"
+    effect = "Deny"
+
+    principals {
+      type        = "*"
+      identifiers = ["*"]
+    }
+
+    actions = [
+      "s3:*"
+    ]
+
+    resources = [
+      aws_s3_bucket.this.arn,
+      "${aws_s3_bucket.this.arn}/*"
+    ]
+
+    condition {
+      test     = "Bool"
+      variable = "aws:SecureTransport"
+      values   = ["false"]
+    }
+  }
+
+  dynamic "statement" {
+    for_each = var.cloudfront_arns
+
+    content {
+      effect = "Allow"
+
+      principals {
+        type        = "Service"
+        identifiers = ["cloudfront.amazonaws.com"]
+      }
+
+      actions = [
+        "s3:GetObject"
+      ]
+
+      resources = [
+        "${aws_s3_bucket.this.arn}/*"
+      ]
+
+      condition {
+        test     = "StringEquals"
+        variable = "aws:SourceArn"
+        values   = [statement.value]
+      }
+    }
+  }
+}
+
+resource "aws_s3_bucket_versioning" "this" {
+  count  = var.enable_versioning ? 1 : 0
+  bucket = aws_s3_bucket.this.id
+
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
```

Comment:
We have built up some other conventions over the years that might be helpful to include here.  Like, aborting multipart uploads, and setting expiration of non-current versions.  I am guessing you aren't including tags since the recommendation should be to set global tags at the provider.
https://github.com/racker/janus-waypoint/blob/24afd14cc161a38676953ef300f4b1b0be28387b/terraform/main.tf#L317-L334

Reply:
I originally intended on limiting scope just to security-ish related things, but I figured out a way to do the lifecycle stuff also without it getting too complicated.